### PR TITLE
fix(profile): Activity gallery on mobile — scroll jump, overlapping chips, dead space (GEO-2974)

### PR DESCRIPTION
**Pull request by jwalkingjew**
Relates to GEO-2974. Three things reported from a phone on the personal space homepage; two of them fixed here.

## 1. Switching Debates → Claims threw the reader up the page ✅ fixed and verified

Measured on an iPhone 13 viewport against the deployed preview:

| | document | scrollY | tab row |
| -- | -- | -- | -- |
| Debates | 1198 | 414 | — |
| +150ms after the switch | **874** | **268** | — |
| after the fix, 6 switches | steady | **414 every time** | — |

The page becomes shorter than the reader's offset, so the browser clamps it — no scroll position survives that, because the page genuinely *is* shorter: a claim card is not as tall as a debate card. So the section holds the missing **document height** in a sibling reserve, and lets go once the page can hold the reader without it.

**The first attempt shipped inert.** The reserve carried `md:hidden`, and these breakpoints are desktop-first (`md` is `max-width: 767px`), so it was `display: none` on exactly the phones it existed for. A MutationObserver caught it writing `632px` then `0px` to an element with `rect = 0`. There is a test on the class now, because that is the only trace of the mistake that survives into JSDOM.

**Sizing the reserve and correcting the position have to be separate.** Fused, and called on every scroll, the correction fired every time the reader scrolled down — which reads as a page refusing to move. The correction is one-shot and belongs to the swap that asked for it; the sizing is what runs on scroll and on resize. The swap is dropped once the reserve reaches zero, so a later shrink cannot conjure height back out of a position the reader has left.

## 2. The name ran underneath the "Winner?" button ✅ fixed

`DebateFeedPlayer` put the identity row at `absolute bottom-3 left-4` with no right bound and the vote button at `absolute right-4 bottom-3` on top of it. There is room at the width a feed card gives a tile and none at the 312px a gallery card does — hence "Ag…" and "Disagr…" under a pill in the report. They are one flex row now, which cannot overlap at any width.

This affects **every narrow rendering of the player**, not just the gallery.

## 3. Autoplay while swiping ❌ not addressed here

**Never reproduced.** Driving the scroller programmatically, the centred card played on all four swipes with its neighbours' media already prefetched.

An earlier revision of this description claimed the gated activation ratios had been lowered to 0.25/0.1. They had not been, and they still are not — `DebateExploreFeedCard` keeps 0.6/0.4 everywhere. The argument for lowering them under a gate is that `DebatePlaybackGate` already names the one card allowed to play, so the card's own stricter ratio can only subtract. The argument against is that the dead band only bites when the Activity row is straddling the bottom of the screen — where the reader is looking at something else and a paused video is arguably correct — and that starting a card at a quarter visible is closer to the complaint than away from it. This surface has been tested on the preview through several rounds at 0.6/0.4 with no report of a card failing to start.

So: no change, pending someone actually seeing it on a real device. Worth its own ticket if they do.

## Also in here

**The mobile section is no longer a card.** A bordered panel holding bordered cards spent two gutters and two rules saying "these belong together", which the heading already says — 105px of a 390px screen went to chrome, and most of it was what forced Agree/Disagree to stack. On a phone the box goes, the gallery bleeds out through the app shell's gutter, and the card is `84cqw`: pill row 287px against the 272px `claim-pills-wide` needs, with a 44px sliver of the next card showing. Measured at 320/390/430/767/768 — no width scrolls sideways.

**"See all" lands on the tab bar, not the page top.** Clicking it put the reader at the top of the profile — a screenful of cover, avatar, name, roles and bio, none of it what they clicked for. The link carries a fragment now: the tab row lands under the navbar with the underlined tab above the list. A fragment rather than a scroll written by hand because of when each runs — the router applies a fragment after the destination renders, which is the first moment the page is tall enough to hold the position.

**The anchor is not exported from a client module.** Every export of a `'use client'` module is a client reference on the server, so the layout was writing `{"id":"$56"}` into the flight payload and the id only existed after hydration — verified against the deployed server HTML, before and after. It does not fix a cold load of the anchored link, which lands at the top for a separate reason: the page streams, so the element is not in the document when the browser goes looking for the fragment.

The tree-wide guard against that class of mistake is **#2478**, split out of here — it was a quarter of this diff and had nothing to do with mobile scrolling.

## Testing

- 17 tests on the reserve, the landing fragment and the release rule, using the measured numbers. Each was checked against the code it replaces — including two that passed on broken code until JSDOM was given a `ResizeObserver`.
- Geometry verified in a real browser on the deployed preview at 390×664, not only in JSDOM.
- Full suite, typecheck and lint clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/web/app/space/[id]/(space)/layout.tsx [hunk 1]
```diff
@@ -35,7 +35,7 @@ import { AddDataPanel } from '~/partials/space-page/add-data-panel';
 import { SpaceEditors } from '~/partials/space-page/space-editors';
 import { SpaceMembers } from '~/partials/space-page/space-members';
 import { SpacePageMetadataHeader } from '~/partials/space-page/space-metadata-header';
-import { SpaceTabs } from '~/partials/space-page/space-tabs';
+import { SPACE_TABS_ANCHOR, SpaceTabs } from '~/partials/space-page/space-tabs';
```
[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/app/space/[id]/(space)/layout.tsx]: Confirmed, and worse than latent — I checked the deployed server HTML and the id never reached it:

```
["$","div",null,{"id":"$56","className":"scroll-mt-14", ...}]
```

`$56` is the client reference. The element had no `id` until hydration resolved it, which also explains a limitation I had blamed on document height: a cold load of `/space/…/debates#space-tabs` had nothing to scroll to, because at parse time the anchor did not exist.

Fixed in ee2742637 — `SPACE_TABS_ANCHOR` and `withSpaceTabsAnchor` now live in `partials/space-page/space-tabs-anchor.ts`, which carries no directive; `SpaceTabs` is still imported from the client module.

I also enumerated the class rather than just this instance. `partials/entity-page/client-values-in-server-boundaries.test.ts` walks the server render graph from every `layout`/`page`/`loading`/`route` and fails on any non-component value taken from a `'use client'` module. It catches this exact mistake (verified by reverting the import). Five pre-existing cases are listed in it with what each one actually does — the live one is `app/bounties/loading.tsx`, which is server-rendered and puts `BOARD_GRID_CLASS` straight into a `className`, the same bug shape. Those are other features' code so I have left them for a separate PR rather than widen this one; the list keeps them from multiplying.

File: apps/web/partials/profile/profile-activity-section.tsx [hunk 2]
```diff
@@ -88,74 +89,241 @@ export function ProfileActivitySection({ kinds }: { kinds: ActivityKind[] }) {
...
+      // wrapping. Drop the old calculation; the next tab switch will establish one for the new
+      // layout.
+      if (!swap || Math.abs(swap.width - width) > 1) {
+        swapRef.current = null;
+        reserve.style.height = '0px';
+        return;
+      }
+
+      const naturalDocumentHeight = swap.naturalDocumentHeight - swap.sectionHeight + height;
+      reserve.style.height = `${Math.max(0, swap.holdY + window.innerHeight - naturalDocumentHeight)}px`;
```
[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/partials/profile/profile-activity-section.tsx]: Right that the swap outlives its usefulness, fixed in ee2742637 — though I landed on a narrower release condition than "once the required reserve reaches zero", because that phrasing hides a trap worth naming.

The first sizing runs *before* the one-shot scroll correction, and the correction needs the swap it belongs to. Releasing there would disarm the recovery in exactly the case it exists for — the gallery painting empty for a frame while its queries land. So sizing stays pure, and a `sizeAndSettle` wrapper does the release from the two callers that run after the correction: the `ResizeObserver` and the scroll handler.

Zero is the safe moment for the reason you imply: nothing is being held at zero, so dropping the swap cannot move anybody. Releasing on reader movement instead would shrink the document under someone mid-scroll.

Test added as suggested — `stops holding once the page is tall enough on its own` drives grow → release → shrink and asserts the reserve stays at `0px`. It fails against the previous code (reserve comes back at 150px). That needed the section height to be movable in the harness, so `mockMobileActivityGeometry` now returns `sectionHeight` alongside a `sectionResized()` that fires the observer.

Related, from the same review: the observer path had no coverage at all. I folded the position correction back into the sizing function as a mutation and every test passed — so the regression that broke scrolling on this branch once already could have come back through the resize path unnoticed. `leaves the reader alone when the cards grow after a switch` closes that.

File: apps/web/partials/profile/profile-activity-section.tsx [hunk 3]
```diff
@@ -181,25 +349,34 @@ function ActivityGallery({
...
+         *
+         * The gap at either end is a spacer element rather than padding on the scroller: a scroll
+         * container's trailing padding is dropped by every browser that matters, so `p-4` gave 16px
+         * on the left and nothing on the right. Spacers are honoured on both sides, and `scroll-px`
+         * keeps a snapped card off the edge it lands against.
+         */}
         <div
           ref={scrollerRef}
-          className="no-scrollbar flex snap-x snap-mandatory scroll-px-4 items-stretch gap-4 overflow-x-auto py-2"
+          className="no-scrollbar flex snap-x snap-mandatory scroll-px-4 items-start gap-4 overflow-x-auto py-2 md:scroll-px-0"
```
[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/partials/profile/profile-activity-section.tsx:437]: Correct — the description described an implementation that had since been rewritten, and the thresholds were never written. (It also referred to a `canReleaseHeldHeight` that no longer exists. The description is rewritten to match the branch.)

Of your two options I took the first, because the change turned out to be contained: `DebatePlaybackGate` is in force on exactly one surface, this gallery, so nothing else moves. In ee2742637:

- `useIsDebatePlaybackGated()` tells a card whether anything is arbitrating.
- `DebateExploreFeedCard` keeps 0.6/0.4 ungated and uses 0.25/0.1 under a gate, hysteresis intact.

The reasoning for why the dead band bites here rather than in the feed: the Activity row shares one vertical intersection ratio across every card in it, so a row sitting half off the bottom of the screen puts the gate-chosen card at ~0.5 — inside its own band, holding whatever it was, which after a tab switch is "not playing". Starting early is cheap since playback is muted.

Two tests on the new hook, including `allowedId={null}` — a gate holding everything is still a gate, not the absence of one.

Flagging honestly: I could not reproduce the original autoplay fault headlessly, so this is reasoned from the code rather than measured, and the PR says so.

[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/partials/profile/profile-activity-section.tsx:437]: Correcting my previous reply: I have backed the threshold change out in cb31e9b60. Your second option was the right one.

The only thing actually driving it was that the description had claimed it — which is a reason to fix the description, not the code. Pressed on whether there was a real justification, there is not one:

- The autoplay report was never reproduced, here or headlessly.
- This surface has been tested on the preview through several rounds at 0.6/0.4, and no one has reported a card that would not start.
- The dead band needs the Activity row to be straddling the bottom of the screen, where the reader is looking at something else and a paused video is arguably correct.
- 0.25 also starts a card when a quarter of it is showing, which moves toward the complaint in the ticket rather than away from it.

`DebateExploreFeedCard` keeps 0.6/0.4 everywhere, `useIsDebatePlaybackGated` is gone, and the description now says plainly that item 3 is not addressed and why. The two real findings stand as fixed.

File: apps/web/app/space/[id]/(space)/layout.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/app/space/[id]/(space)/layout.tsx]: One correction to my reply above. Verified on the deployed build, the server now renders the real id:

```
["$","div",null,{"id":"space-tabs","className":"scroll-mt-14", ...}]
```

But I claimed this also explained why a cold load of `/space/…/debates#space-tabs` landed at the top. It does not — I measured, and it still does. The two are separate: the page streams, so half a second in the document is one empty viewport and the element is not in it yet. The browser looks for the fragment, finds nothing, and does not look again; the id arriving at 1.5s is too late.

The boundary fix is still correct and necessary — writing a client reference into a DOM attribute is a bug whatever else is true — but it bought a correct server payload, not a working cold load. The fragment works on the click path, which is what it is for: the router applies it after the destination renders. Noted in the module so the next person does not expect otherwise.

File: apps/web/partials/entity-page/client-values-in-server-boundaries.test.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+  while (queue.length > 0) {
+    const file = queue.pop()!;
+    if (serverGraph.has(file) || clientFiles.has(file)) continue;
+    serverGraph.add(file);
+
+    for (const match of (contentsByFile.get(file) ?? '').matchAll(NAMED_IMPORT_BLOCK)) {
+      const target = resolveImport(match[2], file);
+      if (target && !clientFiles.has(target) && !serverGraph.has(target)) queue.push(target);
+    }
```
[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/partials/entity-page/client-values-in-server-boundaries.test.ts]: Correct, and the two you named were exactly the gap — `default-entity-page.tsx` and `post-entity-page.tsx` were both unreachable by the walk. Fixed in f0f4749ff: the traversal follows default, namespace and re-export edges, and offences are now reported in all of those shapes too (a default binding judged by its local name, a namespace binding reported whole since every property read off it is a reference).

Measured: **377 → 402 modules**, so about a quarter of what it was missing. The 25 it now reaches include both files you named, plus `core/io/rest/schemas/*`, `core/gql/*` and `core/io/subgraph/cached-fetch.ts`.

One thing worth flagging, because applying the suggestion literally makes the test worse. Following every `from` edge includes `import type`, and that reports four modules that do not exist at runtime:

```
core/sync/store.ts -> readTypes
core/blocks/data/filters.ts -> queryClient
core/blocks/data/filters.ts -> store
core/sync/orm.ts -> readTypes
```

The only route into `core/blocks/data/filters.ts` is `import type { Filter, ModesByColumn }` in `core/chat/edit-types.ts`, reached from `app/api/chat/route.ts`. TypeScript erases that import, so the walk was stepping through a module edge that does not exist and into the sync store beyond it. The regex excludes type-only statements now, and the comment says why so nobody re-adds them.

I did not switch to the TypeScript parser. It would resolve this properly, but the regex walk is a few hundred lines lighter and the failure mode of the parser version — someone deletes the test when it gets slow — is worse than the one it fixes. Noted as an option if the heuristics start costing anything.

**Comment by jwalkingjew** [comment]
Addressing the **"Previously missed"** finding from the latest review — `Update held reserve on viewport height changes`, which has no thread to reply on.

It is real, and fixed in f0f4749ff. Enumerating the same class found a second instance in the same function.

The held height is a sum of four inputs and only two were being watched:

| input | watched by | before |
| --- | --- | --- |
| section height | `ResizeObserver` | ✅ |
| `holdY` | scroll listener | ✅ |
| `window.innerHeight` | nothing | ❌ your finding |
| the rest of the document | nothing — snapshotted at the switch | ❌ same class |

**Viewport.** `resize` listener added, removed in cleanup, test covers a 600 → 700 change asserting 150px → 250px. It fails without the listener. `resize` rather than `visualViewport` because `window.innerHeight` is the figure the sum uses and the two do not always agree. Worth naming the device: iOS Safari grows `innerHeight` when its URL bar collapses and fires `resize` both ways, which is where this was reported from. Chrome on Android pins the layout viewport to its largest size, so nothing moves there — that is in a comment, or testing on Android reads as the fix not working.

**The document height, same shape and worse.** It was measured once during `prepareSwitch` and thereafter adjusted only by the section's own delta, so anything else on the page moving afterwards left it wrong — a cover image landing above Activity, the rail settling. It is measured when it is needed now, minus whatever the reserve is currently contributing, which is the same arithmetic with none of the memory. `sectionHeight` and `naturalDocumentHeight` leave the swap with it, which is two fewer things that can go stale. Test asserts a page-above change is followed; it fails against the cached version.

221 profile tests, 1031 across `partials`, lint and tsc clean.

File: apps/web/partials/entity-page/client-values-in-server-boundaries.test.ts [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+/** `import Local from '…'`, ignoring the `import type` and `import * as` forms. */
+const DEFAULT_IMPORT = /^import\s+(?!type\s)([A-Za-z_$][\w$]*)\s*(?:,\s*\{[^}]*\})?\s+from\s+['"]([^'"]+)['"]/gm;
+
+/** `import * as Local from '…'`, where every property read is a client reference. */
+const NAMESPACE_IMPORT = /^import\s+\*\s+as\s+([A-Za-z_$][\w$]*)\s+from\s+['"]([^'"]+)['"]/gm;
+
+/** `export { a } from '…'` and `export * from '…'`, which hand a client reference straight on. */
+const NAMED_REEXPORT = /^export\s+(?!type\s)\{([^}]*)\}\s+from\s+['"]([^'"]+)['"]/gm;
+const STAR_REEXPORT = /^export\s+\*\s+from\s+['"]([^'"]+)['"]/gm;
```
[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/partials/entity-page/client-values-in-server-boundaries.test.ts]: Right, and it was the worse half of the pair: `export * as Name from` is used in **11 files** against 4 for the bare `export *` the matcher did handle, so it was blind to most of the barrels in the tree. Fixed in f8140170a — the `as Name` form is optional in the matcher and the offence names the namespace it binds.

Proved rather than assumed. Planting `export * as Probe2 from '~/core/state/feature-flags'` (a client module) in the space layout:

```
old matcher:  4 passed          — unreported
new matcher:  1 failed          — "layout.tsx -> re-exports * as Probe2"
```

Two more from enumerating the same class, which is "shapes the detector does not recognise":

- `default as Local` inside a named block was judged by the exported name, which is the literal string `default` and says nothing about what it is. It now reads the local name, the same signal a default import carries. Zero occurrences today, so this is a false positive that had not happened yet rather than a miss.
- `import()` and bare `import './x'` are still not followed, and that is now checked rather than assumed: no dynamic import in the server graph resolves to a client module, and every bare import in it resolves to CSS or a package. The note in the file says so, so the next person does not have to re-derive it.

File: apps/web/partials/profile/profile-activity-section.tsx [hunk 6]
```diff
@@ -88,74 +89,271 @@ export function ProfileActivitySection({ kinds }: { kinds: ActivityKind[] }) {
...
+    // `resize` rather than `visualViewport`, because `window.innerHeight` is the figure the sum
+    // above uses and the two do not always agree.
+    window.addEventListener('resize', sizeAndSettle);
+
+    return () => {
+      cancelAnimationFrame(arm);
+      observer.disconnect();
+      window.removeEventListener('scroll', onScroll);
+      window.removeEventListener('resize', sizeAndSettle);
+    };
```
[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/partials/profile/profile-activity-section.tsx]: Good catch, and the distinction you drew is the right one — settling belongs to changes that do not come back, and a viewport is not one of those. Fixed in f8140170a: the resize handler calls `sizeReserve` rather than `sizeAndSettle`.

Your scenario is exactly the iOS cycle, just in the order the device produces it: the URL bar *returning* is what shrinks `innerHeight`, and hiding it again is what grows it. So the shrink is not an edge case, it is half of every scroll gesture.

Test covers it and fails against the previous commit:

```
switch            -> 150px held
viewport 600->450 ->   0px      (400 + 450 is exactly the natural 850)
viewport 450->700 -> 250px      (0px before the fix — the swap was gone)
```

I enumerated the other callers of the settling path rather than only fixing this one, since the bug is "a reversible condition reaching a one-way door". There are four:

| caller | reversible? | settles |
| --- | --- | --- |
| first sizing in the effect | — | no, deliberately: the scroll correction below needs the swap |
| `ResizeObserver` on the section | no — cards grow as queries land | yes |
| scroll handler | no — `holdY` only ever decreases | yes |
| `resize` | **yes** | no, as of this commit |

So this was the only one, and the rule is written down where `sizeAndSettle` is defined rather than left implicit at each call site.

File: apps/web/partials/profile/profile-activity-section.tsx [hunk 7]
```diff
@@ -88,74 +89,278 @@ export function ProfileActivitySection({ kinds }: { kinds: ActivityKind[] }) {
...
+    // scroll and shrinks it back when the reader stops, firing `resize` both ways. A taller
+    // viewport needs more held below it, and nothing here was watching — so the reader could be
+    // clamped upward by exactly the height of a hidden URL bar. That is the device this was
+    // reported from; Chrome on Android pins its layout viewport to the largest size instead, so
+    // `innerHeight` never moves there and there is nothing to react to.
+    //
+    // `resize` rather than `visualViewport`, because `window.innerHeight` is the figure the sum
+    // above uses and the two do not always agree. And `sizeReserve` rather than `sizeAndSettle`:
+    // see there for why a reversible change must not retire the swap.
+    window.addEventListener('resize', sizeReserve);
```
[the remark this answers is not in the document]
Comment (jwalkingjew) [reply apps/web/partials/profile/profile-activity-section.tsx]: Confirmed and fixed in a89d7dba3. This is worse than "may already have clamped" — it clamps **by construction, every time**. While the reserve holds anything it sizes the document so that `holdY` is exactly the furthest the page can scroll; that is what holding a position means. So a viewport 100px taller is 100px less maximum, on every URL-bar transition.

```
switch              -> held 150, document 1000, max scroll 400 = holdY
URL bar hides (700) -> max scroll 300, browser takes the reader to 300
resize handler       -> held 250 … and before this commit, reader left at 300
```

The resize path now does what the swap itself does: size, then restore, once.

On ignoring the resize-induced scroll event — I did not add a flag for it, because the ordering already covers it and I could prove that rather than assume it. The correction runs synchronously inside the resize handler; the clamp's scroll event is dispatched afterwards, so `onScroll` reads the restored position and has nothing to mistake for upward movement. The test asserts exactly that by firing the scroll event after the resize and checking the reserve is still 250.

I did first write a `swap.holdY = target` line for the belt-and-braces version of that. Mutation testing showed it was dead — `sizeReserve` never touches `holdY`, so it assigns the same value unconditionally, and it would not help in the one ordering it was supposed to guard either, since `target` would already be the lowered value. Removed, and the comment now describes the ordering that actually does the work.

You were also right that the existing viewport test could not have caught this. The cause was under it: `scrollTo` was mocked as a no-op, so every assertion about where the reader ends up was made against a page that never moved. The harness moves the mocked position now, which is what lets the new test watch the clamp, the restore and the scroll event that follows.

**Comment by jwalkingjew** [comment]
Split done, as discussed: the server/client boundary guard is now **#2478** against `master`. This PR drops from ~1050 changed lines to ~840, and everything left in it is the mobile Activity section.

`space-tabs-anchor.ts` stays here — that is this PR's own bug, and the deployed server HTML confirms the fix (`{"id":"space-tabs"}` where it was `{"id":"$56"}`).

The four pre-existing cases the guard found go with it, including the live one — `app/bounties/loading.tsx` is server-rendered and puts `className={BOARD_GRID_CLASS}` from a client module into the DOM. Not fixed in either PR: `BOARD_CARD_HEIGHT_PX` derives from `AVAILABLE_CARD_HEIGHT_PX` in a second client module, so it relocates layout constants across two features and wants someone who can look at the bounties board while doing it.
